### PR TITLE
Direct sphere

### PR DESCRIPTION
**Pull request by AWBroscius**
Polish the class by adding comments, optimizing code performance, and adding testing.

### CODE REVIEW

File: DiverseSelector/distance.py [hunk 1]
```diff
@@ -106,50 +90,29 @@ def pairwise_similarity_bit(feature: np.array, metric: str) -> np.ndarray:
...
+
+    function_dict = {
+        "tanimoto": tanimoto,
+        "modified_tanimoto": modified_tanimoto,
+    }
+
     pair_simi = []
     size = len(feature)
     for i in range(0, size):
         for j in range(i + 1, size):
```
Comment (FanwangM) [inline DiverseSelector/distance.py:104]: Use `itertools` for simplicity.

File: DiverseSelector/distance.py [hunk 2]
```diff
@@ -173,39 +138,18 @@ def tanimoto(a: np.array, b: np.array) -> float:
     return coeff
```
Comment (FanwangM) [inline DiverseSelector/distance.py:138]: Change the above line to 

```
shared = sum(a * b)
coeff = shared / (sum(a ** 2) + sum(b ** 2) - shared)
```
?

File: DiverseSelector/diversity.py [hunk 3]
```diff
@@ -107,25 +99,31 @@ def entropy(x: np.ndarray) -> float:
...
     so we lose any information associated with num in matrix.
     Weidlich, I. E., and Filippov, I. V. (2016)
     Using the Gini coefficient to measure the chemical diversity of small-molecule libraries.
     Journal of Computational Chemistry 37, 2091-2097.
     """
+
+    # convert input matrix to bit matrix
+    y = np.empty(x.shape)
     for i in range(0, len(x)):
         for j in range(0, len(x[0])):
```
Comment (FanwangM) [inline DiverseSelector/diversity.py:112]: Use 'itertools' as a better option?

File: DiverseSelector/diversity.py [hunk 4]
```diff
@@ -135,46 +133,18 @@ def entropy(x: np.ndarray) -> float:
...
-    A Novel Measure for Assessing the Diversity of Compound Databases.
-    Journal of Chemical Information and Modeling 46, 1898-1904.
-    """
-    cs = len(rdFMCS.FindMCS(mols))
-    nc = len(x)
-    sdi = (1 - nearest_average_tanimoto(x)) / (0.8047 - (0.065 * (np.log(nc))))
-    cr = -1 * np.log10(nc / (cs ** 2))
-    edi = (sdi + cr) * 0.7071067811865476
-    edi_scaled = ((np.tanh(edi / 3) + 1) / 2) * 100
-    return edi_scaled
```
Comment (FanwangM) [inline DiverseSelector/diversity.py:169]: Is there any reason to get rid of this index, `explicit_diversity_index`?

File: DiverseSelector/test/test_distance.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    shortest_tani = [(11/19), (11/19)]
+    average = np.average(shortest_tani)
+    assert_equal(nat, average)
+
+
+
+
+
+
+
```
Comment (FanwangM) [inline DiverseSelector/test/test_distance.py:105]: We often just leave one empty line at the end of PY files.

**Review by FanwangM — CHANGES_REQUESTED** [review]
Thanks for the nice work on this.
Only some minor changes.

**Comment by AWBroscius** [comment]
Hi @FanwangM, thank you for leaving comments! I am going to close this pull request however as my branches got mixed up and it has too many commits unrelated to Direct Spheres. I plan on applying your advice to the correct branch, and opening a new pull request that is just focused on Direct Spheres.

File: DiverseSelector/diversity.py (hunk 4 above)
Comment (AWBroscius) [reply DiverseSelector/diversity.py:169]: We wanted to remove RD Kit as a dependency, and as this function depended on it, we decided to remove it.

Comment (FarnazH) [reply DiverseSelector/diversity.py:169]: @FanwangM, to keep this function, the API needs to change. The `mols` are only used to obtain `cs = len(rdFMCS.FindMCS(mols))`, so `cs` can be an argument. @AWBroscius can you give this a try?

Comment (AWBroscius) [reply DiverseSelector/diversity.py:169]: I can certainly go back and change the code to make `cs` an argument. However, I do not know what to write in the documentation for it. In the original paper, `cs` is the number of core structures/common-substructures present in a library of chemical compounds. How would this translate for non-chemical datasets?  Would it make sense to apply this measure to non-chemical datasets? @FarnazH @FanwangM

Comment (FanwangM) [reply DiverseSelector/diversity.py:169]: Thanks for the clarification. I don't think it's applicable for non-chemical data. We should make this clear in the docstring and we are OK. Probably we can put this information into the `Notes` section. @awb9691

**Pull request closed without merging**